modules/gapi/src/streaming/vpl/vpl_source_engine.cpp

```diff
@@ -0,0 +1,164 @@
+                    sess.last_status = MFXVideoCORE_SyncOperation(sess.session, sess.sync, 100);
+                    if (MFX_ERR_NONE == sess.last_status) {
+
+                        DecodeSession& my_sess = static_cast<DecodeSession&>(sess);
+                        on_frame_ready(my_sess);
+                    }
+                } while (sess.last_status == MFX_WRN_IN_EXECUTION);
+            }
+
+            return this->process_error(sess.last_status, static_cast<DecodeSession&>(sess));
+        }
+    );
+}
+
+void VPLDecodeEngine::initialize_session(mfxSession mfx_session,
+                                         DecoderParams&& decoder_param,
+                                         file_ptr&& source_handle)
+{
+    register_session<DecodeSession>(mfx_session,
+                                    std::move(decoder_param.stream),
+                                    std::move(source_handle));
+    acceleration_policy->init(mfx_session);
+}
+
+VPLProcessingEngine::ExecutionStatus VPLDecodeEngine::execute_op(operation_t& op, EngineSession& sess) {
+    return op(sess);
+}
+    
+void VPLDecodeEngine::on_frame_ready(DecodeSession& sess)
+{
+    mfxFrameInfo *info = &sess.dec_surface_out->Info;
+    mfxFrameData *data = &sess.dec_surface_out->Data;
+    size_t w = info->Width;
+    size_t h = info->Height;
+    size_t p = data->Pitch;
+
+    throw std::runtime_error(std::string(__FUNCTION__) + " is not implemented");
+    
+    GAPI_LOG_DEBUG(nullptr, "[" << sess.session << "], surface: " << sess.dec_surface_out <<
+                            ", w: " << w << ", h: " << h << ", p: " << p);
+    const int cols = info->CropW;
+    const int rows = info->CropH;
+
+    switch (info->FourCC) {
+        case MFX_FOURCC_I420: {
+        } break;
+
+        case MFX_FOURCC_NV12: {
+        } break;
+
+        default:
+            GAPI_LOG_WARNING(nullptr, "Unsupported FourCC code: " << info->FourCC << ". Skip");
+            return;
+    }
+
+    ready_frames.push(cv::Mat(cv::Size{rows, cols}, CV_8UC3));
+}
+
+VPLProcessingEngine::ExecutionStatus VPLDecodeEngine::process_error(mfxStatus status, DecodeSession& sess)
+{
+    GAPI_LOG_DEBUG(nullptr, "status: " << status);
+    switch (status) {
+        case MFX_ERR_NONE:
+            return ExecutionStatus::Continue; 
+        case MFX_ERR_MORE_DATA: // The function requires more bitstream at input before decoding can proceed
+            if (!sess.source_handle) {
+                // No more data to drain from decoder, start encode draining mode
+                return ExecutionStatus::Processed;
+            }
+            else
+                return ExecutionStatus::Continue; // read more data
+            break;
+        case MFX_ERR_MORE_SURFACE:
+            // The function requires more frame surface at output before decoding can proceed.
+            // This applies to external memory allocations and should not be expected for
+            // a simple internal allocation case like this
```

Comment:
I'd put some asserts/messages here since it's empty

Reply:
added for `VPLProcessingLegacyEngine` 
I'll keep in mind this comment
Current implementation is not active (falls on Assert during creation)